### PR TITLE
[MAINT] Refactor fieldmap heuristics

### PR DESCRIPTION
**Pull request by oesteban**
Handle all fieldmaps (including syn) together, so they are easier to maintain (thinking of integrating the ME-EPI with fieldmap correction and also of the `--func-only`).

When no fieldmap correction is posible, then a naive workflow connecting inputs and outputs  is set up.

Some other "better practices" are introduced in this PR:

- [x] Try to avoid passing along BIDS layout objects
- [x] Reduce the number of arguments a workflow takes
- [x] Place DataSinks for reportlets closer to their generation, but avoiding passing `reportlets_dir` on (this is done post-hoc, collecting all nodes which name starts with `'ds_report_'` and setting the proper `base_directory` and `source_file` inputs for all at once). Eventually, we will be able to remove the whole workflow of datasinks for reports.

### CODE REVIEW

**Comment by oesteban** [comment]
Closes #1013

File: fmriprep/workflows/fieldmap/base.py [hunk 1]
```diff
@@ -36,48 +36,223 @@
...
+        workflow.connect([
+            (inputnode, outputnode, [('bold_ref', 'bold_ref'),
+                                     ('bold_mask', 'bold_mask'),
+                                     ('bold_ref_brain', 'bold_ref_brain')]),
+        ])
+        return workflow
+
+    # In case there are multiple fieldmaps prefer EPI
+    fmaps.sort(key=lambda fmap: FMAP_PRIORITY[fmap['type']])
+    fmap = fmaps[0]
```
Comment (effigies) [inline fmriprep/workflows/fieldmap/base.py]: I feel like the selection logic is a bit distributed. I'd rather make decisions in one place, and then just work through the consequences. So here, I would look for two things: (1) Which primary SDC should we run? (2) Are we running SyN?

```Python
fmap = min(fmaps, key=lambda...)
run_syn = any(fm['type'] == 'syn' for fm in fmaps)
```

Yeah, it's iterating twice, but at `n < 4`, it doesn't matter. If you want to sort and use `fmaps[0]` and `fmaps[-1]`, that's fine, but I feel like this is slightly clearer in meaning.

File: fmriprep/workflows/fieldmap/base.py [hunk 2]
```diff
@@ -36,48 +36,223 @@
...
+
+    # PEPOLAR path
+    if fmap['type'] == 'epi':
+        setattr(workflow, 'sdc_method', 'PEB/PEPOLAR (phase-encoding based / PE-POLARity)')
```
Comment (effigies) [inline fmriprep/workflows/fieldmap/base.py]:
```Python
workflow.sdc_method = '...'
```

File: fmriprep/workflows/fieldmap/base.py [hunk 3]
```diff
@@ -36,48 +36,223 @@
...
+                     for fmap_ in fmaps if fmap_['type'] == 'epi']
+        sdc_unwarp_wf = init_pepolar_unwarp_wf(
+            bold_meta=bold_meta,
+            epi_fmaps=epi_fmaps,
+            omp_nthreads=omp_nthreads,
+            name='pepolar_unwarp_wf')
+
+    # FIELDMAP path
+    if fmap['type'] in ['fieldmap', 'phasediff']:
+        setattr(workflow, 'sdc_method', 'FMB (%s-based)' % fmap['type'])
```
Comment (effigies) [inline fmriprep/workflows/fieldmap/base.py]:
```Python
workflow.sdc_method = '...'
```

File: fmriprep/workflows/fieldmap/base.py [hunk 4]
```diff
@@ -36,48 +36,223 @@
...
+                ('bold_ref_brain', 'inputnode.in_reference_brain'),
+                ('bold_mask', 'inputnode.in_mask')]),
+            (fmap_estimator_wf, sdc_unwarp_wf, [
+                ('outputnode.fmap', 'inputnode.fmap'),
+                ('outputnode.fmap_ref', 'inputnode.fmap_ref'),
+                ('outputnode.fmap_mask', 'inputnode.fmap_mask')]),
+        ])
+
+    # FIELDMAP-less path
+    if fmaps[-1]['type'] == 'syn':
```
Comment (effigies) [inline fmriprep/workflows/fieldmap/base.py]:
```Python
if run_syn:
```

File: fmriprep/workflows/fieldmap/base.py [hunk 5]
```diff
@@ -36,48 +36,223 @@
...
+            (inputnode, syn_sdc_wf, [
+                ('t1_brain', 'inputnode.t1_brain'),
+                ('t1_2_mni_reverse_transform', 'inputnode.t1_2_mni_reverse_transform'),
+                ('bold_ref', 'inputnode.bold_ref'),
+                ('bold_ref_brain', 'inputnode.bold_ref_brain'),
+                ('template', 'inputnode.template')]),
+        ])
+
+        # XXX Eliminate branch when forcing isn't an option
+        if len(fmaps) == 1:  # --force-syn was called
```
Comment (effigies) [inline fmriprep/workflows/fieldmap/base.py]: I think this is clearer:

```Python
if fmap['type'] != 'syn':
```

File: fmriprep/workflows/fieldmap/base.py [hunk 6]
```diff
@@ -36,48 +36,223 @@
...
+            setattr(workflow, 'sdc_method', 'FLB ("fieldmap-less" based) - SyN')
```
Comment (effigies) [inline fmriprep/workflows/fieldmap/base.py]:
```Python
workflow.sdc_method = '...'
```

Also, I don't know that "fieldmap-less based" makes sense. Maybe "fieldmap-less, SyN-based", or just "fieldmap-less".

File: fmriprep/workflows/fieldmap/pepolar.py [hunk 7]
```diff
@@ -81,28 +77,22 @@ def init_pepolar_unwarp_wf(fmaps, bold_file, omp_nthreads, layout=None,
...
-        if fmap_pe[0] == bold_file_pe[0]:
-            if len(fmap_pe) != len(bold_file_pe):
-                add_list = usable_fieldmaps_opposite_pe
-            else:
-                add_list = usable_fieldmaps_matching_pe
-            add_list.append(fmap)
+    usable_fieldmaps_matching_pe = []
+    usable_fieldmaps_opposite_pe = []
+    for fmap, fmap_pe in epi_fmaps:
+        fmap_pe = fmap_pe
```
Comment (effigies) [inline fmriprep/workflows/fieldmap/pepolar.py]: This line should be dropped.

File: fmriprep/workflows/fieldmap/syn.py [hunk 8]
```diff
@@ -150,46 +145,46 @@ def init_nonlinear_sdc_wf(bold_file, freesurfer, bold2t1w_dof,
...
+            ('output_image', 'inputnode.in_file')]),
+        (unwarp_ref, outputnode, [('output_image', 'out_reference')]),
         (skullstrip_bold_wf, outputnode, [
             ('outputnode.skull_stripped_file', 'out_reference_brain'),
             ('outputnode.mask_file', 'out_mask')]),
-        (syn_rpt, outputnode, [('out_report', 'out_warp_report')])])
+    ])
 
     return workflow
+
```
Comment (effigies) [inline fmriprep/workflows/fieldmap/syn.py]: Also, PEP8: Two blank lines before/after functions.

File: fmriprep/workflows/fieldmap/base.py (hunk 1 above)
Comment (oesteban) [reply fmriprep/workflows/fieldmap/base.py]: done at the only branching where the `run_syn` is tested, thanks!

**Comment by oesteban** [comment]
Testing right now on a dataset without fieldmaps at all.

Also, removed the opaque way of passing the SDC type along as an object attribute. Nipype has means for that :).

File: fmriprep/workflows/fieldmap/base.py (hunk 2 above)
Comment (oesteban) [reply fmriprep/workflows/fieldmap/base.py]: Now it uses the workflow's outputnode

File: fmriprep/workflows/fieldmap/base.py [hunk 9]
```diff
@@ -36,48 +36,223 @@
...
-        fmap
-          The estimated fieldmap itself IN UNITS OF Hz.
-        fmap_ref
-          the anatomical reference for the fieldmap (magnitude image, corrected SEm, etc.)
-        fmap_mask
-          a brain mask for the fieldmap
+    **Inputs**
+        name_source
+            Original BOLD run filename to fetch metadata (TODO: should
+            be replaced with the ``bold_meta`` and ``fmaps`` input)
```
[the remark this answers is not in the document]
Comment (oesteban) [reply fmriprep/workflows/fieldmap/base.py]: While I totally agree with this comment, I ran the extra mile and actually fixed this TODO.

**Comment by oesteban** [comment]
Okay, seems to work fine with and without `--use-syn`/`--force-syn`. Thumbs up @chrisfilo, @effigies ?

**Comment by effigies** [comment]
:+1:

**Pull request merged**